src/game.cpp

```diff
@@ -12592,6 +12592,10 @@ void game::plswim(int x, int y)
         add_msg(_("The water puts out the flames!"));
         u.remove_effect("onfire");
     }
+    if (u.has_effect("onfire")) {
```

Comment:
> Would it still work on NPCs though?

It should. That function is also used by NPCs and it's called for each active NPC in lightmap.cpp.